rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -1012,21 +1029,44 @@ impl MergeInsertJob {
         mut self,
         source: SendableRecordBatchStream,
     ) -> Result<(Arc<Dataset>, MergeStats)> {
+        let start = Instant::now();
         let mut source_iter =
             super::new_source_iter(source, self.params.conflict_retries > 0).await?;
 
         let mut dataset_ref = self.dataset.clone();
         let max_retries = self.params.conflict_retries;
-        let mut backoff = Backoff::default();
+        let mut backoff = SlotBackoff::default();
         while backoff.attempt() <= max_retries {
             let ds = dataset_ref.clone();
-            let (transaction, stats) = self
+            let (transaction, mut stats) = self
                 .clone()
                 .execute_uncommitted_impl(source_iter.next().unwrap())
                 .await?;
+            stats.num_attempts = backoff.attempt() + 1;
             match CommitBuilder::new(ds).execute(transaction).await {
                 Ok(ds) => return Ok((Arc::new(ds), stats)),
                 Err(Error::RetryableCommitConflict { .. }) => {
+                    // Check whether we have exhausted our retries *before*
+                    // we sleep.
+                    if backoff.attempt() >= max_retries {
+                        break;
+                    }
+                    if start.elapsed() > self.params.retry_timeout {
+                        return Err(Error::TooMuchWriteContention {
+                            message: format!(
+                                "Attempted {} times, but failed on retry_timeout of {:.3} seconds.",
+                                backoff.attempt() + 1,
+                                self.params.retry_timeout.as_secs_f32()
+                            ),
+                            location: location!(),
+                        });
+                    }
+                    if backoff.attempt() == 0 {
+                        // We add 10% buffer here, to allow concurrent writes to complete.
+                        // See SlotBackoff implementation for more details on how this works.
+                        backoff = backoff.with_unit((start.elapsed().as_millis() * 11 / 10) as u32);
+                    }
```

Comment:
Interesting, so after the first attempt we look at how long the attempt took and then use that as our slot unit?

Reply:
Exactly. We sort of assume all concurrent operations are taking about the same time.